### PR TITLE
Add multiprocessing in the DPO trainer.

### PR DESCRIPTION
**Pull request by imraviagrawal**
This tiny PR add support for multi process preprocessing in the DPO trainer. 
Added support for num_proc to tokenize the training dataset.

### CODE REVIEW

File: trl/trainer/dpo_trainer.py [hunk 1]
```diff
@@ -162,6 +164,7 @@ def __init__(
         generate_during_eval: bool = False,
         compute_metrics: Optional[Callable[[EvalLoopOutput], Dict]] = None,
         precompute_ref_log_probs: bool = False,
+        dataset_num_proc = None,
```
Comment (younesbelkada) [inline trl/trainer/dpo_trainer.py]:
```suggestion
        dataset_num_proc: int = 2,
```

**Review by younesbelkada — COMMENTED** [review]
Thanks a lot ! Do you think it would make sense to initialize it with a decent value, e.g. 2 ?

File: trl/trainer/dpo_trainer.py [hunk 2]
```diff
@@ -162,6 +164,7 @@ def __init__(
         generate_during_eval: bool = False,
         compute_metrics: Optional[Callable[[EvalLoopOutput], Dict]] = None,
         precompute_ref_log_probs: bool = False,
+        dataset_num_proc: bool = None,
```
Comment (younesbelkada) [inline trl/trainer/dpo_trainer.py:167]:
```suggestion
        dataset_num_proc: int = 2,
```

File: trl/trainer/dpo_trainer.py [hunk 3]
```diff
@@ -122,6 +122,8 @@ class DPOTrainer(Trainer):
         precompute_ref_log_probs (`bool`, defaults to `False`):
             Flag to precompute reference model log probabilities and evaluation datasets. This is useful if you want to train
             without the reference model and reduce the total GPU memory needed.
+        dataset_num_proc (`Optional[int]`):
```
Comment (younesbelkada) [inline trl/trainer/dpo_trainer.py:125]:
```suggestion
        dataset_num_proc (`int`):
```

File: trl/trainer/dpo_trainer.py (hunk 1 above)
Comment (imraviagrawal) [reply trl/trainer/dpo_trainer.py]: thats reasonable

**Review by younesbelkada — COMMENTED** [review]
Actually this looks good, let's make it optional ! 
Would you be happy to add a test here: https://github.com/huggingface/trl/blob/main/tests/test_dpo_trainer.py#L231 just to make sure training runs fine?
Also make sure to run the styling checks `make precommit`

File: trl/trainer/dpo_trainer.py (hunk 1 above)
Comment (younesbelkada) [reply trl/trainer/dpo_trainer.py]: Please see my latest comment, I think it is good to make it configurable and set it to None by default

**Comment by imraviagrawal** [comment]
@younesbelkada take a look at the test case, we also did make precommit

**Review by younesbelkada — APPROVED** [review]
Thanks a lot for your work on this!

**Pull request merged**